### PR TITLE
API for submit + tests

### PR DESCRIPTION
**Pull request by scottfrazer**

### CODE REVIEW

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala [hunk 1]
```diff
@@ -29,7 +35,9 @@ class MockWorkflowManagerActor extends Actor  {
 
   def receive = {
     case SubmitWorkflow(wdl, inputs) =>
-      println("not yet implemented")
+      Future {
+            MockWorkflowManagerActor.submittedWorkflowId
+      }.pipeTo(sender())
```
Comment (mcovarr) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: Why is this creating a `Future` around the ID to then do a `pipeTo`?  Why not just  
```sender ! MockWorkflowManagerActor.submittedWorkflowId```

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala [hunk 2]
```diff
@@ -44,9 +52,34 @@ class MockWorkflowManagerActor extends Actor  {
...
+      |  }
+      |}
+      |
+      |workflow hello {
+      |  call hello
+      |}
+    """.stripMargin
+
+  val HelloInputs: Map[String, String] = Map("hello.hello.addressee" -> "world")
+  val HelloInputsJson : String = HelloInputs.toJson.toString()
```
Comment (mcovarr) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: A lot of the good versions of this Wdl already live in `SampleWdl`

File: src/main/scala/cromwell/webservice/CromwellApiHandler.scala [hunk 3]
```diff
@@ -3,18 +3,18 @@ package cromwell.webservice
...
 import spray.http.StatusCodes
 
 import scala.concurrent.duration._
-import scala.util.{Failure, Success}
+import scala.util.{Try, Failure, Success}
 
 object CromwellApiHandler {
   sealed trait WorkflowManagerMessage
-  case class SubmitWorkflow(wdl: WdlSource, inputs: WorkflowRawInputs) extends WorkflowManagerMessage
+  case class SubmitWorkflow(wdl: String, inputs: Map[String, String]) extends WorkflowManagerMessage
```
Comment (scottfrazer) [inline src/main/scala/cromwell/webservice/CromwellApiHandler.scala]: The old line is the one we should use:

```scala
case class SubmitWorkflow(wdl: WdlSource, inputs: WorkflowRawInputs) extends WorkflowManagerMessage
```

File: src/main/scala/cromwell/webservice/CromwellApiHandler.scala [hunk 4]
```diff
@@ -41,5 +42,20 @@ class CromwellApiHandler(workflowManagerActorRef : ActorRef) extends Actor {
...
           context.parent ! RequestComplete(StatusCodes.InternalServerError, ex.getMessage)
       }
+
+    case SubmitWorkflow(wdl, inputs) =>
+      implicit val timeout = Timeout(2.seconds)
+      val workflowManagerResponseFuture = ask(workflowManagerActorRef, WorkflowManagerActor.SubmitWorkflow(wdl, inputs)).mapTo[WorkflowId]
+
+      workflowManagerResponseFuture.onComplete {
+        case Success(id) =>
+          println(s"Got in here with $id")
```
Comment (scottfrazer) [inline src/main/scala/cromwell/webservice/CromwellApiHandler.scala]: Remove println (or replace with write to log file)

File: src/main/scala/cromwell/webservice/CromwellApiService.scala [hunk 5]
```diff
@@ -75,4 +80,39 @@ trait CromwellApiService extends HttpService with PerRequestCreator  {
...
+
+                  tryMap match {
+                    case Success(workflowInputsMap) =>
+                      requestContext =>
+                        println("WI:" +  workflowInputsMap.toString)
+                        perRequest(requestContext, CromwellApiHandler.props(workflowManagerActorRef), CromwellApiHandler.SubmitWorkflow(wdlSource, workflowInputsMap))
+                    case Failure(ex) =>
+                      complete(StatusCodes.BadRequest, "workflowInput JSON was malformed")
+                  }
+                }
```
Comment (scottfrazer) [inline src/main/scala/cromwell/webservice/CromwellApiService.scala]: Indenting is off on this file... `Code -> Reformat Code...`

File: src/main/scala/cromwell/webservice/CromwellApiService.scala [hunk 6]
```diff
@@ -75,4 +80,39 @@ trait CromwellApiService extends HttpService with PerRequestCreator  {
...
+    path("workflows") {
+        post {
+                formFields('wdlSource.as[String], 'workflowInputs.as[String]) { (wdlSource, workflowInputs) =>
+
+                  val tryMap = Try(workflowInputs.parseJson.convertTo[Map[String, String]])
+
+                  tryMap match {
+                    case Success(workflowInputsMap) =>
+                      requestContext =>
+                        println("WI:" +  workflowInputsMap.toString)
```
Comment (scottfrazer) [inline src/main/scala/cromwell/webservice/CromwellApiService.scala]: Remove println

File: src/main/scala/cromwell/webservice/CromwellApiHandler.scala [hunk 7]
```diff
@@ -41,5 +42,20 @@ class CromwellApiHandler(workflowManagerActorRef : ActorRef) extends Actor {
...
+          context.parent ! RequestComplete (StatusCodes.Created, WorkflowSubmitResponse(id.toString, engine.WorkflowSubmitted.toString))
+
+        case Failure(ex) =>
+          println(s"Failure2 -> $ex")
```
Comment (scottfrazer) [inline src/main/scala/cromwell/webservice/CromwellApiHandler.scala]: Remove println

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala (hunk 1 above)
Comment (kcibul) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: now I know;)

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala (hunk 2 above)
Comment (kcibul) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: check

File: src/main/scala/cromwell/webservice/CromwellApiHandler.scala (hunk 3 above)
Comment (kcibul) [inline src/main/scala/cromwell/webservice/CromwellApiHandler.scala]: check

File: src/main/scala/cromwell/webservice/CromwellApiService.scala [hunk 8]
```diff
@@ -49,16 +52,16 @@ class CromwellApiServiceActor(val workflowManagerActorRef : ActorRef, swaggerSer
...
-  val workflowRoutes = queryRoute
+  val workflowRoutes = queryRoute ~ submitRoute
 
   @ApiOperation(value = "Query for workflow based on workflow id.",
     nickname = "query",
     httpMethod = "GET",
-    produces = "application/json",
-    response = classOf[WorkflowStatusResponse]
+    produces = "application/json"
+    //, response = classOf[WorkflowStatusResponse]
```
Comment (kcibul) [inline src/main/scala/cromwell/webservice/CromwellApiService.scala]: This (and the next one) are the lines that were causing all the strange compiler problems.  Not sure how to fix.  The impact right now is the swagger doesn't describe the output message format.

File: src/main/scala/cromwell/webservice/CromwellApiService.scala [hunk 9]
```diff
@@ -75,4 +78,38 @@ trait CromwellApiService extends HttpService with PerRequestCreator  {
           complete(StatusCodes.BadRequest)
       }
     }
+
+
+  @ApiOperation(value = "Submit a new workflow for execution",
+    nickname = "submit",
+    httpMethod = "POST",
+    produces = "application/json"
+//  ,response = classOf[WorkflowSubmitResponse]
```
Comment (kcibul) [inline src/main/scala/cromwell/webservice/CromwellApiService.scala]: The second occurrence

**Comment by scottfrazer** [comment]
Looks good to me.  I tried to make some sense of this compiler error this morning.

One thing to note is that `sbt compile` does work, but it's the assembly that seems to be creating the issues.  Judging from the output of `sbt assembly`, I think perhaps it could be a conflict with another library, because it seems to have the error immediately after importing a bunch of JARs:

```
...
[info] Including: jackson-jaxrs-json-provider-2.4.1.jar
[info] Including: jackson-module-jsonSchema-2.4.1.jar
[info] Including: jackson-jaxrs-base-2.4.1.jar
[error] missing or invalid dependency detected while loading class file 'WorkflowStatusResponse.class'.
[error] Could not access type AnyRef in package scala,
[error] because it (or its dependencies) are missing. Check your build definition for
[error] missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
[error] A full rebuild may help if 'WorkflowStatusResponse.class' was compiled against an incompatible version of scala.
[error] missing or invalid dependency detected while loading class file 'WorkflowSubmitResponse.class'.
[error] Could not access type AnyRef in package scala,
[error] because it (or its dependencies) are missing. Check your build definition for
[error] missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
[error] A full rebuild may help if 'WorkflowSubmitResponse.class' was compiled against an incompatible version of scala.
[error] two errors found
[error] (test:compileIncremental) Compilation failed
[error] Total time: 32 s, completed Jun 2, 2015 8:39:34 AM
```

**Comment by scottfrazer** [comment]
I even tried the `-Ylog-classpath` and got this output: https://gist.github.com/scottfrazer/7c0336c521df59a896a0

However, this is quite cryptic and I haven't been able to make enough sense of it to debug the issue.  I also just for the sake of it, tried `import scala.AnyRef` because... why not?  I tried it in a 2 different places that the compiler was complaining about and it didn't make a difference.

**Comment by scottfrazer** [comment]
So I decided I'd personally go look for where `AnyRef` was defined... so I went into my `~/.ivy/cache` directory and found all Scala jars and sources I could  with a script and then did a `jar -tf` grepping for AnyRef.  [The results](https://gist.github.com/scottfrazer/6274b7a77028818ca018) were that I can't find it anywhere!!  This is very odd because things compile for me on other branches.  I guess I'm not looking in the right place or something.

Even removing my Ivy cache and redownloading *everything*, I still could not find AnyRef defined anywhere.

My goal was to find which JAR AnyRef was defined in and then make sure that classpath was included, which would surprise me if it weren't because it's fine on any other branch.

File: src/main/scala/cromwell/webservice/CromwellApiService.scala [hunk 10]
```diff
@@ -75,4 +78,38 @@ trait CromwellApiService extends HttpService with PerRequestCreator  {
...
+    new ApiResponse(code = 201, message = "Successful Request"),
+    new ApiResponse(code = 400, message = "Malformed Input"),
+    new ApiResponse(code = 500, message = "Internal Error")
+  ))
+  def submitRoute =
+    path("workflows") {
+      post {
+        formFields("wdlSource", "workflowInputs") { (wdlSource, workflowInputs) =>
+
+          val tryMap = Try(workflowInputs.parseJson.convertTo[Map[String, String]])
```
Comment (mcovarr) [inline src/main/scala/cromwell/webservice/CromwellApiService.scala]: Must the JSON be parseable as `Map[String, String]` or would we allow boolean or numeric values?

Comment (scottfrazer) [inline src/main/scala/cromwell/webservice/CromwellApiService.scala]: I'll take care of this with the merge into my PR, which will handle this a little better.

**Comment by pgrosu** [comment]
@scottfrazer You can't find `AnyRef` since its not defined as a class, and thus using it would need to be compiled with your `scala-library.jar` file.  Below is an example:
```
$ cat ExtAnyRef.scala
class ExtAnyRef extends AnyRef { }
object ExtAnyRef { }
$ 
$ cat AnyRef.java
class AnyRef extends ExtAnyRef {
  public static void main(String[] args) {
        System.out.println("Hello World!");
    }
}
$
$ scalac -cp . ExtAnyRef.scala
$ javac  -cp .:$SCALA_HOME/lib/scala-library.jar AnyRef.java
$ java   -cp .:$SCALA_HOME/lib/scala-library.jar AnyRef
Hello World!
$
```

Below is a link that describes it nicely:

http://stackoverflow.com/questions/2335319/what-are-the-relationships-between-any-anyval-anyref-object-and-how-do-they-m

Hope it helps,
~p

**Comment by scottfrazer** [comment]
Interesting, I thought something funny like that was happening.  Good to know!

**Comment by pgrosu** [comment]
Gladly :)

**Pull request merged**